src/Compiler/VMCode.idr

```diff
@@ -101,59 +106,113 @@ Show VMInst where
   show (MkVMFun args body) = show args ++ ": " ++ show body
+  show (MkVMForeign ccs args ret)
+      = "Foreign call " ++ show ccs ++ " " ++
+        show args ++ " " ++ show ret
   show (MkVMError err) = "Error: " ++ show err
 
 toReg : AVar -> Reg
 toReg (ALocal i) = Loc i
 toReg ANull = Discard
 
-toVM : (tailpos : Bool) -> (target : Reg) -> ANF -> List VMInst
-toVM t Discard _ = []
+0
+TagM : Type -> Type
+TagM = State (Int, NameMap Int)
+
+initState : (Int, NameMap Int)
+initState = (0, empty)
+
+runTagM : TagM a -> a
+runTagM = evalState initState
+
+getTag : Name -> TagM Int
+getTag n = do
+    nm <- gets snd
+    let Nothing = lookup n nm
+        | Just t => pure t
+    t <- gets fst
+    put (t + 1, insert n t nm)
+    pure t
+
+projectArgs : Int -> Int -> (used : IntMap ()) -> (args : List Int) -> List VMInst
+projectArgs scr i used [] = []
+projectArgs scr i used (arg :: args)
+    = case lookup arg used of
+           Just _ => PROJECT (Loc arg) (Loc scr) i :: projectArgs scr (i + 1) used args
+           Nothing => projectArgs scr (i + 1) used args
+
+collectReg : Reg -> IntMap ()
+collectReg (Loc i) = singleton i ()
+collectReg _ = empty
+
+collectUsed : VMInst -> IntMap ()
+collectUsed (DECLARE reg) = collectReg reg
+collectUsed START = empty
+collectUsed (ASSIGN _ val) = collectReg val
+collectUsed (MKCON _ _ args) = foldMap collectReg args
+collectUsed (MKCLOSURE _ _ _ args) = foldMap collectReg args
+collectUsed (MKCONSTANT _ _) = empty
+collectUsed (APPLY _ fn arg) = collectReg fn <+> collectReg arg
+collectUsed (CALL _ _ _ args) = foldMap collectReg args
+collectUsed (OP _ _ args) = foldMap collectReg args
+collectUsed (EXTPRIM _ _ args) = foldMap collectReg args
+collectUsed (CASE _ is mdef)
+    = foldMap (foldMap collectUsed . snd) is
+      <+> maybe empty (foldMap collectUsed) mdef
+collectUsed (CONSTCASE _ is mdef)
+    = foldMap (foldMap collectUsed . snd) is
+      <+> maybe empty (foldMap collectUsed) mdef
+collectUsed (PROJECT _ val _) = collectReg val
+collectUsed (NULL _) = empty
+collectUsed (ERROR _) = empty
+
+toVM : (tailpos : Bool) -> (target : Reg) -> ANF -> TagM (List VMInst)
+toVM t Discard _ = pure []
 toVM t res (AV fc (ALocal i))
-    = [ASSIGN res (Loc i)]
+    = pure [ASSIGN res (Loc i)]
 toVM t res (AAppName fc _ n args)
-    = [CALL res t n (map toReg args)]
+    = pure [CALL res t n (map toReg args)]
 toVM t res (AUnderApp fc n m args)
-    = [MKCLOSURE res n m (map toReg args)]
+    = pure [MKCLOSURE res n m (map toReg args)]
 toVM t res (AApp fc _ f a)
-    = [APPLY res (toReg f) (toReg a)]
+    = pure [APPLY res (toReg f) (toReg a)]
 toVM t res (ALet fc var val body)
-    = toVM False (Loc var) val ++ toVM t res body
+    = [| toVM False (Loc var) val ++ toVM t res body |]
 toVM t res (ACon fc n ci (Just tag) args)
-    = [MKCON res (Left tag) (map toReg args)]
+    = pure [MKCON res tag (map toReg args)]
 toVM t res (ACon fc n ci Nothing args)
-    = [MKCON res (Right n) (map toReg args)]
```

Comment:
Wouldn't it be better to this work on a more high level IR so that the `ACon`
already arrives with an `Int` tag no matter whether the constructor is a type
or data one? This way all of the backends could benefit from this optimisation!

Reply:
Yep, I'll do that!